src/extension-consumer/src/produce/mod.rs

```diff
@@ -1,149 +1,269 @@
+    #[structopt(short, long)]
+    pub files: Vec<PathBuf>,
+}
+
+fn validate_key_separator(separator: String) -> std::result::Result<(), String> {
+    if separator.is_empty() {
+        return Err(
+            "must be non-empty. If using '=', type it as '--key-separator \"=\"'".to_string(),
+        );
+    }
+    Ok(())
 }
 
 impl ProduceLogOpt {
-    pub async fn process(self, fluvio: &Fluvio) -> Result<()> {
-        let (cfg, file_records) = self.validate()?;
-        let mut producer = fluvio.topic_producer(&cfg.topic).await?;
+    pub async fn process(mut self, fluvio: &Fluvio) -> Result<()> {
+        let mut producer = fluvio.topic_producer(&self.topic).await?;
 
-        if let Some(records) = file_records {
-            produce_from_files(&mut producer, cfg, records).await?;
+        // --key-separator implies --lines
+        if self.key_separator.is_some() {
+            self.lines = true;
+        }
+
+        if self.files.is_empty() {
+            self.lines = true;
+            if atty::is(atty::Stream::Stdin) {
+                eprintln!("Reading one record per line from stdin:");
+            }
+            self.produce_stdin(&mut producer).await?;
         } else {
-            produce_stdin(&mut producer, cfg).await?;
+            self.produce_from_files(&mut producer).await?;
         }
 
         Ok(())
     }
 
-    /// Validate cli options. Generate target-server and produce log configuration.
-    pub fn validate(self) -> Result<(ProduceLogConfig, Option<FileRecord>)> {
-        let file_records = if let Some(record_per_line) = self.record_per_line {
-            Some(FileRecord::Lines(record_per_line))
-        } else if !self.record_file.is_empty() {
-            Some(FileRecord::Files(self.record_file))
-        } else {
-            None
-        };
+    /// Sends records to a Topic based on the file configuration given
+    ///
+    /// This will either send the lines of a single file as individual records,
+    /// or it will send the entirety of a list of files as records, where each
+    /// whole file is one record.
+    async fn produce_from_files(&self, producer: &mut TopicProducer) -> Result<()> {
+        for path in &self.files {
+            if self.lines {
+                self.produce_file_lines(producer, path).await?;
+            } else {
+                self.produce_file_whole(producer, path).await?;
+            }
+            print_cli_ok!();
+        }
 
-        let produce_log_cfg = ProduceLogConfig {
-            topic: self.topic,
-            partition: self.partition,
-            continuous: self.continuous,
-        };
+        Ok(())
+    }
 
-        Ok((produce_log_cfg, file_records))
+    /// Sends each line from a file as a unique record to the producer's topic.
+    async fn produce_file_lines(&self, producer: &mut TopicProducer, path: &Path) -> Result<()> {
+        let file = File::open(path)?;
```

Comment:
Hm I am reading some of your other comments and I think I see what you're getting at. I will see if I can condense some of this. I left it separate previously because I had originally wanted to print status messages that were unique to the mode being used (e.g. print file with line number, etc.). But I will take a look at this now.

Reply:
file is input (it implements async `read` and `write`), there is no need to differentiate from stdin or any other input.  So we should have a single pipeline for producer just that source is different

Reply:
Ok, I consolidated the implementations, things are a bit cleaner now. I also remembered why I split it the way that I did: it was because there were slightly different strategies for printing acknowledgement to the user based on where the input was coming from. For example, reading from stdin would cause the producer to print `Ok!` after each line was typed (to notify user), but reading from files would only cause producer to print `Ok!` once at the end of the file. I think I have figured out a good middle ground to have both.

Reply:
Looks good